ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -582,6 +578,23 @@ private BlockImportResult importBlockAndState(
     return result;
   }
 
+  private boolean applyProposerBoost(
+      final SignedBeaconBlock block, final StoreTransaction transaction) {
+    if (spec.getCurrentSlot(transaction).equals(block.getSlot())) {
+      final UInt64 millisPerSlot = spec.getMillisPerSlot(block.getSlot());
+      final UInt64 timeIntoSlotMillis = getMillisIntoSlot(transaction, millisPerSlot);
+
+      if (isBeforeAttestingInterval(millisPerSlot, timeIntoSlotMillis)) {
+        if (forkChoiceProposerBoostUniquenessEnabled) {
+          // is_first_block
+          return recentChainData.getStore().getProposerBoostRoot().isEmpty();
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
```

Comment:
NIT: This is a suggestion on how to rewrite this logic w/o the nested if statements. I know that not everyone is a fan of multiple exit points in a method but this makes it easier to add/remove rules to the proposer boost logic (and to me also clearly shows the distinct rules).

```suggestion
  private boolean applyProposerBoost(
      final SignedBeaconBlock block, final StoreTransaction transaction) {
    if (!spec.getCurrentSlot(transaction).equals(block.getSlot())) {
      return false;
    }

    final UInt64 millisPerSlot = spec.getMillisPerSlot(block.getSlot());
    final UInt64 timeIntoSlotMillis = getMillisIntoSlot(transaction, millisPerSlot);
    if (!isBeforeAttestingInterval(millisPerSlot, timeIntoSlotMillis)) {
      return false;
    }

    final boolean isFirstBlock = recentChainData.getStore().getProposerBoostRoot().isEmpty();
    if (forkChoiceProposerBoostUniquenessEnabled && !isFirstBlock) {
      return false;
    }

    return true;
  }
```

Potentially, we could move `isFirstBlock` into the if statement to avoid evaluating it w/o the need to do so, for the sake of readability. But the toggle will be removed soon so we will need to perform this check all the time anyway.

Reply:
Here is another suggestion if we want to stay closer to the spec:
```
  private boolean applyProposerBoost(
      final SignedBeaconBlock block, final StoreTransaction transaction) {
    final UInt64 millisPerSlot = spec.getMillisPerSlot(block.getSlot());
    final UInt64 timeIntoSlotMillis = getMillisIntoSlot(transaction, millisPerSlot);

    final boolean isBlockForCurrentSlot = spec.getCurrentSlot(transaction).equals(block.getSlot());
    final boolean isBlockBeforeAttestingInterval = isBeforeAttestingInterval(millisPerSlot, timeIntoSlotMillis);
    final boolean isFirstBlock =
        forkChoiceUpdateHeadOnBlockImportEnabled ? recentChainData.getStore().getProposerBoostRoot().isEmpty() : true;

    return isBlockForCurrentSlot && isBlockBeforeAttestingInterval && isFirstBlock;
  }
```
I don't love the ternary condition there, but it will probably be removed once we don't need that booster flag anymore.